packages/plugin-react/src/index.ts

```diff
@@ -0,0 +1,258 @@
+import babel, { ParserOptions, TransformOptions, types as t } from '@babel/core'
+import { createFilter } from '@rollup/pluginutils'
+import resolve from 'resolve'
+import type { Plugin, PluginOption } from 'vite'
+import {
+  addRefreshWrapper,
+  isRefreshBoundary,
+  preambleCode,
+  runtimeCode,
+  runtimePublicPath
+} from './fast-refresh'
+import { babelImportToRequire } from './jsx-runtime/babel-import-to-require'
+import { restoreJSX } from './jsx-runtime/restore-jsx'
+
+export interface Options {
+  include?: string | RegExp | Array<string | RegExp>
+  exclude?: string | RegExp | Array<string | RegExp>
+  /**
+   * Enable `react-refresh` integration. Vite disables this in prod env or build mode.
+   * @default true
+   */
+  fastRefresh?: boolean
+  /**
+   * Set this to `"automatic"` to use [vite-react-jsx](https://github.com/alloc/vite-react-jsx).
+   * @default "classic"
+   */
+  jsxRuntime?: 'classic' | 'automatic'
+  /**
+   * Babel configuration applied in both dev and prod.
+   */
+  babel?: TransformOptions
+  /**
+   * @deprecated Use `babel.parserOpts.plugins` instead
+   */
+  parserPlugins?: ParserOptions['plugins']
+}
+
+export default function viteReact(opts: Options = {}): PluginOption[] {
+  let base = '/'
+  let projectRoot = process.cwd()
+  let isProduction = true
+  let skipFastRefresh = opts.fastRefresh === false
+  let filter = createFilter(opts.include, opts.exclude)
+
+  const userPlugins = opts.babel?.plugins || []
+  const userParserPlugins =
+    opts.parserPlugins || opts.babel?.parserOpts?.plugins || []
+
+  const viteBabel: Plugin = {
+    name: 'vite:babel',
+    enforce: 'pre',
+    async transform(code, id, ssr) {
+      if (/\.[tj]sx?$/.test(id)) {
```

Comment:
I know there is a pattern people write JSX in `.js` but does it will happen on `.ts`?

Reply:
Sorry. I found this is a vite react plugin not a vite react refresh plugin. It could be also used as a `babel-loader` .